packages/core/src/components/panel-stack/panelStack.tsx

```diff
@@ -110,25 +116,35 @@ export class PanelStack extends AbstractPureComponent2<IPanelStackProps, IPanelS
         }
     }
 
-    private renderCurrentPanel() {
-        const { showPanelHeader = true } = this.props;
+    private renderPanels() {
+        const { renderCurrentPanelOnly = true } = this.props;
         const { stack } = this.state;
         if (stack.length === 0) {
             return null;
         }
-        const [activePanel, previousPanel] = stack;
+        const panelsToRender = renderCurrentPanelOnly ? [stack[0]] : stack;
+        const panelViews = panelsToRender.map(this.renderPanel).reverse();
+        return panelViews;
+    }
+
+    private renderPanel = (panel: IPanel, index: number) => {
+        const { showPanelHeader = true } = this.props;
+        const { stack } = this.state;
+        const active = index === 0;
+        const layer = stack.length - index;
+        const key = `${layer}-${active}`;
```

Comment:
The `active` part is required for `CSSTransition` to properly animate the bottom panel to move in while _popping_ and moving out while _pushing_. It changes the `key` though, so the `PanelView` is rerendered anyway, so no state is preserved. Removing it makes it work, the state is preserved, but the bottom panel is not animated at all.

I'm not sure how to tackle this issue.

Reply:
I'm confused, what does key have to do with CSSTransition? I don't see it in the docs http://reactcommunity.org/react-transition-group/css-transition

Reply:
The thing is that with `renderCurrentPanelOnly={false}` we would keep all the `CSSTransition`s rendered, therefore they would not trigger the enter event as they were _entered_. To force the enter event, I wanted to change the `key` (just like the current implementation), but `stack.length` is no longer OK, as the length changes with the, well, stack length, and a single panel should not rerender as long as it's hidden.

This `key` contains two parts: first one, `stack.length - index` is constant (and unique) for each panel, second one, `active` changes only when the panel becomes or stops being active.